File: u1_4/implEuler.c

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

typedef struct Result {
	size_t length;
	double* times;
	double* values;
} Result;

double newtonSchritt(double x0, double f, double fAbleitung){
	return x0 - f/fAbleitung;
}

#define MAX(a,b) (a > b ? a : b)
#define errorFunction(yi) (yAbleitungsGleichung(yi,tn) - (yi-yn)/h)

Result implizitesEulerVerfahren(double (*yAbleitungsGleichung)(double y, double t), double y0, double t0, double tEnd, double h){
	
	size_t length = (size_t) MAX(((tEnd-t0)/h+1), 1);
	double* times = (double*) malloc(sizeof(double) * length);
	double* values = (double*) malloc(sizeof(double) * length);
	
	double yn = y0;
	
	double maxError = 1e-3 * h;
	
	times[0] = t0;
	values[0] = y0;
	
	for(size_t i=1;i<length;i++){
		
		double tn = t0 + h * i;
		
		// newtonSchritt
		int maxSteps = 5000, stepCount = 0;
		double yn_p1 = yn, error;
		do {
			
			// f ist die Errorfunktion...
			double f0 = errorFunction(yn_p1);
			double f1 = errorFunction(yn_p1+h);
			double dfdx = (f1-f0)/h;
			double xn_p1 = newtonSchritt(yn_p1, f0, dfdx);
			error = abs(xn_p1-yn_p1);
			yn_p1 = xn_p1;
			stepCount++;
			
		} while(error > maxError && stepCount < maxSteps);
		
		times[i] = tn;
		values[i] = yn_p1;
		
		yn = yn_p1;
		
	}
	
	Result result;
	result.length = length;
	result.times = times;
	result.values = values;
	return result;
	
}

double funcB(double y, double t){
	return 2*t*exp(y)/(1+t*t);
}

void main(){
	
	for(int i=1;i<=8;i++){
		double h = pow(10, -i);
		Result iev = implizitesEulerVerfahren(funcB, 0.0, -2.0, 0.0, h);
		// todo print it or export it
		printf("10^-%d: %.17g, target: %.17g\n", i, iev.values[iev.length-1], -log(1.0 + log(5.0)));
	}
	
}

File: u1_4/implEuler.r

implizitesEulerVerfahren = function(yAbleitungsGleichung, y0, t0, tEnd, h) {
  
  # y(t+h)-y(t)
  # ----------- = y'
  #      h
  
  yn = y0
  times = seq(t0, tEnd, h)
  values = rep(0, length(times))
  maxError = 1e-3 * h
  
  for (index in 1:length(times)) {
    tn = times[index]
    
    errorFunction = function(yi) {
      return(yAbleitungsGleichung(yi) - (yi - yn) / h)
    }
    
    yn_p1 = findeNullstelle(yn,
                            errorFunction,
                            createDerivative(errorFunction, h),
                            maxError)
    yn = yn_p1
    
    values[index] = yn
    
  }
  
  plot(times, values)
  
  
}

# sucht die Nullstelle von f per Newton-Verfahren
findeNullstelle = function(x0, fVonX, fAbleitungVonX, maxError) {
  error = maxError * 2
  maxSteps = 5000
  stepCount = 0
  xn = x0
  while (error > maxError && stepCount < maxSteps) {
    xn_p1 = newtonSchritt(xn, fVonX, fAbleitungVonX)
    error = abs(xn_p1 - xn)
    xn = xn_p1
    stepCount = stepCount + 1
  }
  return(xn)
}

# x1
newtonSchritt = function(x0, fVonX, fAbleitungVonX) {
  return(x0 - fVonX(x0) / fAbleitungVonX(x0))
}

createDerivative = function(fVonX, h) {
  return(function(x)
    (fVonX(x + h) - fVonX(x)) / h)
}

# test: Pendel, weil wir wissen, was rauskommt?
# y' = -c * x
implizitesEulerVerfahren(function(x)
  sqrt(max(0, 1-x*x)), 0.5, 0, 1, 1e-2)
